### PR TITLE
Fs cloning api handler

### PR DESCRIPTION
**Pull request by mpacer**
This adds the handler needed to get file system cloning to work. 

I will create a separate PR just for adding the handler to the routes and testing the routes. 

**NB**: This is a bit of a bigger PR, but I didn't want to submit a PR for functionality without tests for that functionality.

### CODE REVIEW

File: bookstore/clone.py [hunk 1]
```diff
@@ -406,3 +406,101 @@ def construct_template_params(self, relpath, fs_clonepath):
     def get_template(self, name):
         """Loads a Jinja template by name."""
         return BOOKSTORE_FILE_LOADER.load(self.settings['jinja2_env'], name)
+
+
+class BookstoreFSCloneAPIHandler(APIHandler):
+    """Handle notebook clone from a local file system.
```
Comment (willingc) [inline bookstore/clone.py]:
```suggestion
    """Handle notebook clone from an accessible file system (local or cloud).
```

File: bookstore/clone.py [hunk 2]
```diff
@@ -406,3 +406,101 @@ def construct_template_params(self, relpath, fs_clonepath):
...
+
+    Provides API handling for ``POST`` and clones a notebook
+    from the local file system.
```
Comment (willingc) [inline bookstore/clone.py]:
```suggestion
    from the specified file system (local or cloud).
```

File: bookstore/clone.py [hunk 3]
```diff
@@ -406,3 +406,101 @@ def construct_template_params(self, relpath, fs_clonepath):
...
+
+    Methods
+    -------
+    initialize(self)
+        Helper to access bookstore settings.
+    post(self)
+        Clone a notebook from the location specified by the payload.
```
Comment (willingc) [inline bookstore/clone.py]:
```suggestion
        Clone a notebook from the filesystem location specified by the payload.
```

File: bookstore/clone.py [hunk 4]
```diff
@@ -406,3 +406,101 @@ def construct_template_params(self, relpath, fs_clonepath):
...
+    --------
+    `Jupyter Notebook reference on Custom Handlers <https://jupyter-notebook.readthedocs.io/en/stable/extending/handlers.html#registering-custom-handlers>`_
+    """
+
+    def initialize(self):
+        """Helper to retrieve bookstore setting for the session."""
+        self.bookstore_settings = BookstoreSettings(config=self.config)
+
+    def _get_content(self, path):
+        """Helper for getting content from a local filepath.
```
Comment (willingc) [inline bookstore/clone.py]:
```suggestion
        """Helper for getting content from a specified filepath.
```

File: bookstore/clone.py [hunk 5]
```diff
@@ -406,3 +406,101 @@ def construct_template_params(self, relpath, fs_clonepath):
...
+        
+        Parameters
+        ----------
+        path : str
+            Source path from which to get a notebook
```
Comment (willingc) [inline bookstore/clone.py]:
```suggestion
           File system path to a notebook
```

File: bookstore/clone.py [hunk 6]
```diff
@@ -406,3 +406,101 @@ def construct_template_params(self, relpath, fs_clonepath):
...
+    async def post(self):
+        """POST /api/bookstore/fs-clone
+
+        Clone a notebook to the path specified in the payload.
+
+        The payload type for the request should be::
+
+            {
+            "relpath": string,
+            "target_path"?: string
```
Comment (willingc) [inline bookstore/clone.py]:
```suggestion
            "target_path": string
```

Comment (willingc) [reply bookstore/clone.py]: Typo?

File: bookstore/clone.py [hunk 7]
```diff
@@ -270,7 +270,7 @@ def build_content_model(self, content, target_path):
         dict
             Jupyter Contents API compatible model
         """
-        path = self.contents_manager.increment_filename(target_path)
+        path = self.contents_manager.increment_filename(target_path, insert='-')
```
Comment (mpacer) [inline bookstore/clone.py]: To bring @chyzzqo2's comment & response over here:

I added this to ensure consistency between the incrementing patterns for s3 and fs cloning.

File: bookstore/clone.py [hunk 8]
```diff
@@ -406,3 +406,101 @@ def construct_template_params(self, relpath, fs_clonepath):
...
+        ----------
+        content : dict or string
+            dict or string encoded file content
+        target_path : str
+            The the path we wish to clone to, may be incremented if already present.
+
+        Returns
+        --------
+        dict
+            Jupyter Contents API compatible model
```
Comment (willingc) [inline bookstore/clone.py]: Link to Contents API https://jupyter-notebook.readthedocs.io/en/stable/extending/contents.html

File: bookstore/clone.py (hunk 6 above)
Comment (mpacer) [reply bookstore/clone.py]: It's an optional parameter and I was using the typescript convention of adding a `?` after a key for describing the payload type.

Comment (willingc) [reply bookstore/clone.py]: I would leave off the ? and indicate that target_path is optional. As a Python file, I wouldn't expect the typescript convention to be used.

File: bookstore/clone.py [hunk 9]
```diff
@@ -406,3 +406,101 @@ def construct_template_params(self, relpath, fs_clonepath):
...
+    async def post(self):
+        """POST /api/bookstore/fs-clone
+
+        Clone a notebook to the path specified in the payload.
+
+        The payload type for the request should be::
+
+            {
+            "relpath": string,
+            "target_path": string #optional
```
Comment (mpacer) [inline bookstore/clone.py]: Does this successfully capture the idea in a way that would make sense for a python developer?

Comment (mpacer) [inline bookstore/clone.py]: @willingc 
Does this successfully capture the idea in a way that would make sense from a python perspective?

**Pull request merged**